qutebrowser/browser/greasemonkey.py

```diff
@@ -0,0 +1,289 @@
+
+    def meta_json(self):
+        return json.dumps({
+            'name': self._name,
+            'description': self._description,
+            'matches': self._includes,
+            'includes': self._includes,
+            'excludes': self._excludes,
+            'run-at': self._run_at,
+        })
+
+    def meta_raw(self):
+        pass
+
+
+class GreasemonkeyManager(QObject):
+
+    """Manager of userscripts and a greasemonkey compatible environment.
+
+    Signals:
+        scripts_reloaded: Emitted when scripts are reloaded from disk.
+            Any any cached or already-injected scripts should be
+            considered obselete.
+    """
+
+    scripts_reloaded = pyqtSignal()
+
+    def __init__(self):
+        super().__init__()
+        self.load_scripts()
+
+    @cmdutils.register(name='greasemonkey-reload',
+                       instance='greasemonkey')
+    def load_scripts(self):
+        """Re-Read greasemonkey scripts from disk."""
+        self._run_start = []
+        self._run_end = []
+        self._run_idle = []
+
+        scripts_dir = QDir(_scripts_dir())
+        log.greasemonkey.debug("Reading scripts from: %s",
+                               scripts_dir.absolutePath())
+        for script_filename in scripts_dir.entryList(['*.js'], QDir.Files):
+            script_path = scripts_dir.absoluteFilePath(script_filename)
+            log.greasemonkey.debug("Trying to load script: %s", script_path)
+            with open(script_path, encoding='utf-8') as script_file:
+                try:
+                    script = GreasemonkeyScript.parse(script_file.read())
+                except Exception as err:
+                    log.greasemonkey.warning("Unable to load: %s %s",
+                                             script_path, err)
+                    # Catch-all :/
+                    continue
+                if script.run_at() == 'document-start':
+                    self._run_start.append(script)
+                elif script.run_at() == 'document-end':
+                    self._run_end.append(script)
+                elif script.run_at() == 'document-idle':
+                    self._run_idle.append(script)
+                else:
+                    log.greasemonkey.warning(("Script %s has invalid run-at "
+                                              "defined, defaulting to "
+                                              "document-end"), script_path)
+                    # Default as per
+                    # https://wiki.greasespot.net/Metadata_Block#.40run-at
+                    self._run_end.append(script)
+                log.greasemonkey.debug("Loaded script: %s", script.name())
+        self.scripts_reloaded.emit()
+
+    def scripts_for(self, url):
+        """Fetch scripts that a registered to run for url.
+
+        returns a tuple of lists of scripts meant to run at (document-start,
+        document-end, document-idle)
```

Comment:
I think it'd be nicer to define a small [attr class](http://www.attrs.org/en/stable/) like:

```python
@attr.s
class MatchingScripts:

    start = attr.ib()
    end = attr.ib()
    idle = attr.ib()
```

Then from the outside, you can do `scripts = foo.scripts_for(...)` and then access `scripts.start` etc. instead of relying on the position.